File: play.py
###Initilization
import tellurium as te
import csv
from matplotlib import pyplot as plt



ini1 = '''
   A -> B; kfAB*A;
   B -> A; kbAB*B;
   
   A -> C; kfAC*A;
   C -> A; kbAC*C;
   
   B -> D; kfBD*B;
   D -> B; kbBD*D;
   
   C -> D; kfCD*C;
   D -> C; kbCD*D;
   
   B -> C; kfBC*B;
   C -> B; kbBC*C; 
 
   A = 5; B = 5; C = 5; D = 5;
           
   at(time > 10) : A = 10;
   
'''

ini2='''

   kfAB=5; kbAB=2;
   kfAC=1; kbAC=2;
   kfBD=5; kbBD=3;
   kfCD=3; kbCD=1;
   kfBC=10;kbBC=10;
   
'''

###Simulation

r = te.loada(ini1+ini2)
result = r.simulate (0,20,10000)
r.plot() #Just to have the general idea

###Output the plot, results, and the setup info.
plt.savefig("concentrations_vs_time.png")

with open('simulation_result.csv', 'w') as csvfile:
  writer = csv.writer(csvfile)
  [writer.writerow(r) for r in result]

f=open("simulation_setup.txt","w")
f.write(ini2)
f.close()

File: play.R
###Initilization
library(dplyr)
library(ggplot2)

options(max.print=3000)
graphics.off()
par(mar=c(1,1,1,1))

###Load simualtion data
dat=read.csv("simulation_result.csv",header = F)
colnames(dat) = c("time","A","B","C","D")
head(dat) #Just to get the general idea
dat=as_tibble(dat)

###Load setup data
eval(parse("simulation_setup.txt"))

###Data wrangling
###Calculate Rate of the reactions
dat=mutate(dat, RfAB = kfAB * A, RbAB = kbAB * B, 
            RfAC = kfAC * A, RbAC = kbAC * C,
            RfBD = kfBD * B, RbBD = kbBD * D,
            RfCD = kfCD * C, RbCD = kbCD * D,
            RfBC = kfBC * B, RbBC = kbBC * C)

###Calculate Flow
flow=mutate(dat,FAB=(RfAB-RbAB),
                FAC=(RfAC-RbAC),
                FBD=(RfBD-RbBD),
                FCD=(RfCD-RbCD),
                FBC=(RfBC-RbBC))
flow=select(flow,time,FAB,FAC,FBD,FCD,FBC)

###Select the disruption time
flow=filter(flow,time>10)
flow=filter(flow,time<15)

###Output plots as pdf. Specially Look at the last plot.
pdf("results.pdf",width = 16,height = 9)
plot(flow$time,flow$FAB)
plot(flow$time,flow$FAC)
plot(flow$time,flow$FBD)
plot(flow$time,flow$FCD)
plot(flow$time,flow$FBC)
dev.off()
